keps/sig-auth/20190607-certificates-api.md

```diff
@@ -0,0 +1,285 @@
+
+// These are the possible conditions for a certificate request.
+const (
+  CertificateApproved RequestConditionType = "Approved"
+  CertificateDenied   RequestConditionType = "Denied"
+)
+```
+
+### Automatic CSR Approval
+
+The kube-controller-manager ships with an in-built
+[approver](https://github.com/kubernetes/kubernetes/blob/32ec6c212ec9415f604ffc1f4c1f29b782968ff1/pkg/controller/certificates/approver/sarapprove.go)
+for Kubelet TLS Bootstrap that delegates various permissions on CSRs for node
+credentials to authorization. It does this by posting subject access reviews to
+the API server. It punts on TLS certificates for server authentication of the
+Kubelet API because verifying IP SANs for Kubelets in a generic way poses
+challenges.
+
+The GCP controller manager replaces the in-built approver with an
+[approver](https://github.com/kubernetes/cloud-provider-gcp/blob/08fa1e3260ffb267682762e24ba93692000e3be8/cmd/gcp-controller-manager/csr_approver.go)
+that handles TLS certificates for server authentication and support for
+verification of CSRs with proof of control of GCE vTPM.
+
+An external project, [kapprover](https://github.com/coreos/kapprover), does
+policy based approval of kubelet CSRs.
+
+<!--
+TODO: various things use the certificates API for stuff other than Kubelet TLS
+certficates. For example:
+
+* Istio
+* CockroachDB
+
+Investigate these usages and report back.
+-->
+
+### Manual CSR Approval
+
+A Kubernetes administrator (with appropriate permissions) can manually approve
+(or deny) Certificate Signing Requests by using the `kubectl certificate
+approve` and `kubectl certificate deny` commands.
+
+### Automatic Signers
+
+The kube-controller-manager ships with an in-built
+[signer](https://github.com/kubernetes/kubernetes/blob/32ec6c212ec9415f604ffc1f4c1f29b782968ff1/pkg/controller/certificates/signer/cfssl_signer.go)
+that signs all approved certificates with a local key. This key is generally
+configured to be the key of the cluster's root certificate authority, but is not
+required to be.
+
+The GCP controller manager implements a
+[signer](https://github.com/kubernetes/cloud-provider-gcp/blob/08fa1e3260ffb267682762e24ba93692000e3be8/cmd/gcp-controller-manager/csr_signer.go)
+that uses a webhook to sign all approved CSRs. This allows the root certificate
+authority secret material to be stored and maintained outside of the Kubernetes
+control plane.
+
+<!--
+### Certificates API Clients
+
+TODO: talk about what's provided by k8s.io/client-go and what other people have
+built, e.g.
+[certificate-init-container](https://github.com/kelseyhightower/certificate-init-container).
+-->
+
+## Graduation Criteria
+
+### Beta -> GA Graduation
+
+- TBD
+
+## Implementation History
+
+- The Certificates API was merged as Alpha in 1.4
+- The Certificates API was promoted to Beta in 1.6
+
+## Known Issues
+
+- API shape/issues
+  - Requires requesters to know all the info about the end certificate.
```

Comment:
Can you provide an example of when this is a problem?

Reply:
1. It's often duplicated work. If an approver needs to recollect the information to make an authorizing decision, this work by the client is unnecessary and could be replaced with a motivation (e.g. "I want a serving cert valid for the services that front me").
1. It complicates authorization. To use the certificates API to distribute serving certificates, users need to grant pods permission to query services in some capacity.